miden-crypto/src/word.rs

```diff
@@ -0,0 +1,616 @@
+use alloc::string::String;
+use core::{
+    cmp::Ordering,
+    fmt::Display,
+    hash::{Hash, Hasher},
+    ops::Deref,
+    slice,
+};
+
+use thiserror::Error;
+use winter_crypto::Digest;
+
+const WORD_SIZE_FELT: usize = 4;
+const WORD_SIZE_BYTES: usize = 32;
+
+use super::{Felt, StarkField, ZERO};
+use crate::{
+    rand::Randomizable,
+    utils::{
+        ByteReader, ByteWriter, Deserializable, DeserializationError, HexParseError, Serializable,
+        bytes_to_hex_string, hex_to_bytes,
+    },
+};
+
+// DIGEST TRAIT IMPLEMENTATIONS
+// ================================================================================================
+
+/// A unit of data consisting of 4 field elements.
+#[derive(Debug, Default, Copy, Clone, Eq, PartialEq)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+#[cfg_attr(feature = "serde", serde(into = "String", try_from = "&str"))]
+pub struct Word([Felt; WORD_SIZE_FELT]);
+
+impl Word {
+    /// The serialized size of the digest in bytes.
+    pub const SERIALIZED_SIZE: usize = WORD_SIZE_BYTES;
+
+    /// Creates a new [Word] from the given field elements.
+    pub const fn new(value: [Felt; WORD_SIZE_FELT]) -> Self {
+        Self(value)
+    }
+
+    /// Returns the digest as a slice of field elements.
+    pub fn as_elements(&self) -> &[Felt] {
+        self.as_ref()
+    }
+
+    /// Returns the digest as a byte array.
+    pub fn as_bytes(&self) -> [u8; WORD_SIZE_BYTES] {
+        <Self as Digest>::as_bytes(self)
+    }
+
+    /// Returns an iterator over the elements of multiple digests.
+    pub fn digests_as_elements_iter<'a, I>(digests: I) -> impl Iterator<Item = &'a Felt>
+    where
+        I: Iterator<Item = &'a Self>,
+    {
+        digests.flat_map(|d| d.0.iter())
+    }
+
+    /// Returns all elements of multiple digests as a slice.
+    pub fn digests_as_elements(digests: &[Self]) -> &[Felt] {
+        let p = digests.as_ptr();
+        let len = digests.len() * WORD_SIZE_FELT;
+        unsafe { slice::from_raw_parts(p as *const Felt, len) }
+    }
```

Comment:
Together, these are used three times across the codebase, and I'm assuming probably not outside this crate. So can we move those where they are used or at least make them `pub(crate)` instead?

I would also replace "digest" with "word" in the method names.

Reply:
I've updated all the ones I think make sense to update